salt/config/__init__.py

```diff
@@ -840,6 +840,13 @@ def _gather_buffer_space():
     # Always generate minion id in lowercase.
     'minion_id_lowercase': bool,
 
+    # The behavior when generating a minion ID. Can specify 'str' or 'func'.
+    # If 'str' is specified (the default), the 'id' option should be set to
+    # the minion ID or left unspecified for default minion ID generation
+    # behavior. If 'func' is specified, the 'id' option should be set to an
+    # exec module function to run to determine the minion ID.
+    'minion_id_type': str,
```

Comment:
Uggh, I'd not go there, because [YAGNI](https://en.wikipedia.org/wiki/You_aren't_gonna_need_it). If one needs to take something from a file, then just:
```python
def foobar(*args, **kwargs):
    return open(kwargs.get('filename')).read()
```
And then your config would be:
```yaml
minion_id_function:
    module.foobar:
        filename: /etc/hostname
```
As an example.